org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/UsagesCommand.java

```diff
@@ -0,0 +1,197 @@
+import org.eclipse.jdt.core.ISourceReference;
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.search.IJavaSearchConstants;
+import org.eclipse.jdt.core.search.IJavaSearchScope;
+import org.eclipse.jdt.core.search.SearchEngine;
+import org.eclipse.jdt.core.search.SearchMatch;
+import org.eclipse.jdt.core.search.SearchParticipant;
+import org.eclipse.jdt.core.search.SearchPattern;
+import org.eclipse.jdt.core.search.SearchRequestor;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.ResourceUtils;
+import org.eclipse.jdt.ls.core.internal.hover.JavaElementLabels;
+import org.eclipse.lsp4j.Location;
+import org.eclipse.lsp4j.SymbolKind;
+import org.eclipse.lsp4j.TextDocumentPositionParams;
+
+public class UsagesCommand {
+  private static final Set<Integer> INTERESTING_ELEMENT_TYPES;
+
+  static {
+    INTERESTING_ELEMENT_TYPES = new HashSet<>();
+    INTERESTING_ELEMENT_TYPES.add(IJavaElement.METHOD);
+    INTERESTING_ELEMENT_TYPES.add(IJavaElement.INITIALIZER);
+    INTERESTING_ELEMENT_TYPES.add(IJavaElement.TYPE);
+    INTERESTING_ELEMENT_TYPES.add(IJavaElement.CLASS_FILE);
+    INTERESTING_ELEMENT_TYPES.add(IJavaElement.COMPILATION_UNIT);
+    INTERESTING_ELEMENT_TYPES.add(IJavaElement.PACKAGE_FRAGMENT);
+  }
+
+  public static List<UsagesResponse> execute(List<Object> parameters, IProgressMonitor pm) {
+    TextDocumentPositionParams param =
+        JavaModelUtil.convertCommandParameter(parameters.get(0), TextDocumentPositionParams.class);
+    SearchEngine engine = new SearchEngine();
+
+    Map<IJavaElement, SearchResult> results = new HashMap<IJavaElement, SearchResult>();
+    try {
+      ITypeRoot typeRoot = JDTUtils.resolveTypeRoot(param.getTextDocument().getUri());
+      IJavaElement elementToSearch =
+          JDTUtils.findElementAtSelection(
+              typeRoot,
+              param.getPosition().getLine(),
+              param.getPosition().getCharacter(),
+              JavaLanguageServerPlugin.getPreferencesManager(),
+              pm);
+
+      if (elementToSearch == null) {
+        return Collections.emptyList();
+      }
+      IJavaProject[] projects =
+          JavaCore.create(ResourcesPlugin.getWorkspace().getRoot()).getJavaProjects();
+      int flags = IJavaSearchScope.SOURCES | IJavaSearchScope.APPLICATION_LIBRARIES;
+      if (isInsideJRE(typeRoot)) {
+        flags |= IJavaSearchScope.SYSTEM_LIBRARIES;
+      }
+      IJavaSearchScope scope = SearchEngine.createJavaSearchScope(projects, flags);
+
+      SearchPattern pattern =
+          SearchPattern.createPattern(elementToSearch, IJavaSearchConstants.REFERENCES);
+      engine.search(
+          pattern,
+          new SearchParticipant[] {SearchEngine.getDefaultSearchParticipant()},
+          scope,
+          new SearchRequestor() {
```

Comment:
I guess it is deserved to be an inner class.

Reply:
Yes, it's quite long.